packages/studio/src/store/slices/pages/createPageSlice.ts

```diff
@@ -158,6 +163,102 @@ export const createPageSlice: SliceCreator<PageSlice> = (set, get) => {
         (component) => component.uuid === activeComponentUUID
       );
     },
+    getFirstSelectedComponentUUID: () => {
+      const { selectedComponentUUIDs, getActivePageState } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (!componentTree) {
+        throw new Error(
+          `Error rendering components: component tree is undefined.`
+        );
+      }
+      const orderedSelectedComponentUUIDs =
+        ComponentTreeHelpers.mapComponentTreeParentsFirst(
+          componentTree,
+          (c) => {
+            if (selectedComponentUUIDs.has(c.uuid)) {
+              return c.uuid;
+            }
+            return undefined;
+          }
+        ).filter((uuid) => !!uuid);
+      return orderedSelectedComponentUUIDs[0];
+    },
+    addSelectedComponentUUID: (selectedUUID: string) => {
+      set((store) => {
+        store.selectedComponentUUIDs.add(selectedUUID);
+      });
+    },
+    addSelectedComponentRect: (
+      selectedUUID: string,
+      rect: DOMRectProperties
+    ) => {
+      set((store) => {
+        store.selectedComponentRectsMap[selectedUUID] = rect;
+      });
+    },
+    clearSelectedComponents: () => {
+      set((store) => {
+        store.selectedComponentUUIDs = new Set();
+        store.selectedComponentRectsMap = {};
+      });
+    },
+    /**
+     * Adds all of the components and their children ordered between the selected component and
+     * the active component (inclusive) to selectedComponentUUIDs.
+     */
+    addShiftSelectedComponentUUIDs: (selectedComponent: ComponentState) => {
+      const selectedUUID = selectedComponent.uuid;
+      const {
+        activeComponentUUID,
+        getActivePageState,
+        addSelectedComponentUUID,
+        clearSelectedComponents,
+      } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (selectedUUID === activeComponentUUID) {
```

Comment:
nit: can this conditional be moved after the `!activeComponentUUID` check? I think it reads a little easier that way

Reply:
Done